<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function annotates a result from a test cohort with information from the PGRM"><title>Annotate a result set with the PGRM — annotate_results • pgrm</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Annotate a result set with the PGRM — annotate_results"><meta property="og:description" content="This function annotates a result from a test cohort with information from the PGRM"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pgrm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/hugheylab/pgrm/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Annotate a result set with the PGRM</h1>
      <small class="dont-index">Source: <a href="https://github.com/hugheylab/pgrm/blob/HEAD/R/pgrm.R" class="external-link"><code>R/pgrm.R</code></a></small>
      <div class="d-none name"><code>annotate_results.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function annotates a result from a test cohort with information from the PGRM</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">annotate_results</span><span class="op">(</span></span>
<span>  <span class="va">results</span>,</span>
<span>  use_allele_dir <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ancestry <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  build <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  phecode_version <span class="op">=</span> <span class="st">"V1.2"</span>,</span>
<span>  calculate_power <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  annotate_CI_overlap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  LOUD <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>results</dt>
<dd><p>A data frame (or data.table) with results of a test cohort; columns
for SNP, phecode, cases, controls, odds_ratio, P (see demo files for example (e.g results_MGI))</p></dd>


<dt>use_allele_dir</dt>
<dd><p>If TRUE, direction of effect is used when assessing if an
association is replicated. To use this argument, odds ratios must be reported
for the alternative allele</p></dd>


<dt>ancestry</dt>
<dd><p>A string that specifies ancestry of the PGRM that is then used
to annotate the results file.
Options EAS, EUR, AFR, SAS, AMR, ALL. Default ALL</p></dd>


<dt>build</dt>
<dd><p>A string indicating the genome reference build used in the results
table. Options hg19, hg38. Default is hg19.</p></dd>


<dt>phecode_version</dt>
<dd><p>A string indicating the phecode version used in the results
table. (Currently only V1.2 is supported)</p></dd>


<dt>calculate_power</dt>
<dd><p>If TRUE then power calculations will be conducted using
case and control counts from the results file. Necessary for get_AER(). Default FALSE</p></dd>


<dt>annotate_CI_overlap</dt>
<dd><p>If TRUE then a column called "annotate_CI_overlap"
is added to the table, values:
<strong>overlap</strong>: 95% CIs of PGRM and test cohort overlap
<strong>test_cohort_greater</strong>: 95% CI of test cohort greater than PGRM
<strong>PGRM_greater</strong>: 95% CI of PGRM greater than test cohort
If annotate_CI_overlap is TRUE, then results must include 95% CIs</p></dd>


<dt>LOUD</dt>
<dd><p>If TRUE then progress info is printed to the terminal. Default FALSE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.table of the results file annotated with columsn from the PGRM</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes a dataframe with summary statistics from a test cohort.
For an example of the way to format the results data frame, see one of the results
sets included in the package (e.g. results_MGI). (NOTE: If the direction of effect
is used to determine if an association is replicated, then the odds ratios of
the result set must be oriented to the alternative allele.)</p>
<p>The function returns a data.table with the following annotations:</p><ul><li><p>Phecode informtion, including phecode_string and phecode_category</p></li>
<li><p>Allele frequencies from GnomAD (column AF), ancestry specified by the ancestry argument</p></li>
<li><p>The rsID</p></li>
<li><p>The direction of effect (ref or alt) and risk allele of the original association</p></li>
<li><p>Summary statistics from the GWAS catalog association, including the -log10(P),
odds ratio, and 95% confidence intervals (cat_LOG10_P, cat_OR, cat_L95, cat_U95)</p></li>
<li><p>The study accession ID from the GWAS catalog</p></li>
<li><p>A column called powered, which is 1 or 0 indicating whether the test association
is powered &gt; 80% (1 if cases &gt;= cases_needed)</p></li>
<li><p>A column called rep that indicates if the association is replicated (i.e.
p&lt;0.05 in the test cohort; if use_allele_dir==TRUE, then the direction of effect
from the test cohort must also be consistant with what is reported in the catalog)</p></li>
<li><p>If annotate_CI_overlap is true, then information about the relationship
between the 95% CIs from the catalog and the test set is included in column
CI_overlap, and new columns for odds_ratio, L95, and U95 are created(rOR, rL95,
rU95) that are oriented to the risk allele. (This option requires that the confidence
intervals are reported in the test cohort summary statistics)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Example result sets: <a href="results_BBJ.html">results_BBJ</a>, <a href="results_UKBB.html">results_UKBB</a>, <a href="results_BioVU_EUR.html">results_BioVU_EUR</a>, <a href="results_BioVU_AFR.html">results_BioVU_AFR</a>, <a href="results_MGI.html">results_MGI</a></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://pgrm.hugheylab.org">pgrm</a></span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## annotate the BioVU African ancestry result set</span></span></span>
<span class="r-in"><span><span class="va">anno</span> <span class="op">=</span> <span class="fu">annotate_results</span><span class="op">(</span><span class="va">results_BioVU_AFR</span>,ancestry<span class="op">=</span><span class="st">'AFR'</span>, build<span class="op">=</span><span class="st">'hg19'</span>,calculate_power<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Doing power calculations"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get the replication rate of associations powered at &gt;80%</span></span></span>
<span class="r-in"><span><span class="fu"><a href="get_RR.html">get_RR</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Replicated 9 of 11 for RR=81.8%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.8181818</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get the replication rate of all associations</span></span></span>
<span class="r-in"><span><span class="fu"><a href="get_RR.html">get_RR</a></span><span class="op">(</span><span class="va">anno</span>,include<span class="op">=</span><span class="st">'all'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Replicated 17 of 31 for RR=54.8%</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.5483871</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Get the actual:expected ratio</span></span></span>
<span class="r-in"><span><span class="fu"><a href="get_AER.html">get_AER</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Expected 20.2, replicated 17 for AE=0.843 (31</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> associations for 14 uniq phecodes)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.8429951</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Lisa Bastarache, Jake Hughey.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

