<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="pgrm">
<title>Running the PGRM on summary statistics • pgrm</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running the PGRM on summary statistics">
<meta property="og:description" content="pgrm">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">pgrm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/my-vignette.html">Running the PGRM on summary statistics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/hugheylab/pgrm/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Running the PGRM on summary statistics</h1>
            
            <h4 data-toc-skip class="date">2022-11-24</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/hugheylab/pgrm/blob/HEAD/vignettes/my-vignette.Rmd" class="external-link"><code>vignettes/my-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>my-vignette.Rmd</code></div>
    </div>

    
    
<p>The Phenotype Genotype Reference Map (PGRM) is a set of associations
from the (<a href="https://www.ebi.ac.uk/gwas/" class="external-link">NHGRI-EBI GWAS
catalog</a>. The pgrm package enables users to calculate replication
related measures on test cohorts.</p>
<p>This vignette shows how replication measures can be calculated based
on summary statistics from a test cohort. (To generate summary
statistics from raw data, see the run_PGRM_assoc() function.)</p>
<div class="section level2">
<h2 id="load-packages">Load packages<a class="anchor" aria-label="anchor" href="#load-packages"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://pgrm.hugheylab.org">'pgrm'</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://r-datatable.com" class="external-link">'data.table'</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-gwas-summary-statistics-for-test-cohort">Load GWAS summary statistics for test cohort<a class="anchor" aria-label="anchor" href="#load-gwas-summary-statistics-for-test-cohort"></a>
</h2>
<p>The first step is to load in a file of summary statistics. This file
must include columns for the SNP, phecode, cases, controls, odds ratios,
and P-values. 95% confidence intervals are optional but recommended.
SNPs must be denoted by a string with the chromosome, position,
reference and alternate allele, separated by “:”. Builds hg19 and hg37
are both supported. Odds ratios and 95% upper/lower confidence invervals
(U95, L95) should be oriented to the reference allele.</p>
<p>In the example below, we open a csv file containing summary
statistics (in this case, the summary statistics from the BioVU cohort)
using the fread() function. Alternatively you can read your datafile in
as a data.frame and convert it to a data.table using the function
data.table().</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">d</span><span class="op">=</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span></span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'data_sample.csv'</span>, package <span class="op">=</span> <span class="st">'pgrm'</span><span class="op">)</span>,</span>
<span>   colClasses <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>character <span class="op">=</span> <span class="st">'phecode'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">d</span><span class="op">)</span></span>
<span><span class="co">#&gt;             SNP phecode cases controls odds_ratio            P       L95</span></span>
<span><span class="co">#&gt; 1: 1:894573:G:A     157   727    53432  0.9010491 0.2277126520 0.7607171</span></span>
<span><span class="co">#&gt; 2: 2:422144:T:C   250.2 11263    40224  0.9625318 0.3089630031 0.8942630</span></span>
<span><span class="co">#&gt; 3: 2:653575:C:T   250.2 11263    40224  0.9299909 0.0005285958 0.8925921</span></span>
<span><span class="co">#&gt; 4: 4:744972:G:T   250.2 11263    40224  1.0732594 0.0789713391 0.9918522</span></span>
<span><span class="co">#&gt; 5: 4:858332:C:T     332   679    46106  1.2214965 0.0218357270 1.0295019</span></span>
<span><span class="co">#&gt; 6: 4:939087:C:T     332   679    46106  1.2407473 0.0002761054 1.1045725</span></span>
<span><span class="co">#&gt;          U95</span></span>
<span><span class="co">#&gt; 1: 1.0672685</span></span>
<span><span class="co">#&gt; 2: 1.0360124</span></span>
<span><span class="co">#&gt; 3: 0.9689566</span></span>
<span><span class="co">#&gt; 4: 1.1613483</span></span>
<span><span class="co">#&gt; 5: 1.4492968</span></span>
<span><span class="co">#&gt; 6: 1.3937101</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="annotate-test-chort-summary-statistics">Annotate test chort summary statistics<a class="anchor" aria-label="anchor" href="#annotate-test-chort-summary-statistics"></a>
</h2>
<p>Next, you can add annotations from the PGRM to the test cohort
results with the annotate_results() function. Use the build argument to
specify the build of your SNP labels (hg19 or hg38). The ancestry
argument should be set to the genetic ancestry of the cohort (EAS, EUR,
AFR, SAS, AMR, and ALL).</p>
<p>The annotate_results() funciton will annotate all SNP/phenotype pairs
that are present in both the summary statistics table and the PGRM, and
returns a data.table of annotated summary statistics. The annotation
includes information about power, replication, and the source study of
the association. A boolean variable <code>powered</code> indicates if
the association is powered at 80% (given the reported odds ratio and
allele frequency, along with the case and control counts from the test
cohort). A boolean variable <code>rep</code> indicates if the
association is replicated in the test cohort at p&lt;0.05 and the
direction of effect is consistent in the source and test summary
statistics. To match on the direction of effect, the odds ratios in the
test cohort must be oriented to the alternate allle. If your data is not
orginalized in this way, this functionality can be suppressed with the
opion use_allele_dir==FALSE.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate_results.html">annotate_results</a></span><span class="op">(</span><span class="va">d</span>, ancestry <span class="op">=</span> <span class="st">'EUR'</span>, build <span class="op">=</span> <span class="st">'hg19'</span>, calculate_power <span class="op">=</span> <span class="cn">TRUE</span>, LOUD <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span>
<span><span class="co">#&gt;    assoc_ID              SNP phecode cases controls odds_ratio            P</span></span>
<span><span class="co">#&gt; 1:    29441 10:100702737:G:A     340  3385    47361  0.9940978 8.983416e-01</span></span>
<span><span class="co">#&gt; 2:     6712 10:101283330:C:A     244  9059    47425  0.9603257 1.625522e-02</span></span>
<span><span class="co">#&gt; 3:    21791 10:101284237:T:G     555  2440    31734  0.8681993 2.756970e-06</span></span>
<span><span class="co">#&gt; 4:   252232 10:101284237:T:G   555.1  2028    31734  0.8640033 9.246056e-06</span></span>
<span><span class="co">#&gt; 5:   304319 10:101284237:T:G   555.2  1156    31734  0.8843484 3.918215e-03</span></span>
<span><span class="co">#&gt; 6:   284865 10:101287764:G:A   555.1  2028    31734  0.8643269 9.639947e-06</span></span>
<span><span class="co">#&gt;          L95       U95 ancestry       rsID risk_allele_dir    RAF</span></span>
<span><span class="co">#&gt; 1: 0.9077962 1.0886039      EUR rs12260159             ref 0.9200</span></span>
<span><span class="co">#&gt; 2: 0.9291361 0.9925622      EUR rs10748781             ref 0.4208</span></span>
<span><span class="co">#&gt; 3: 0.8183860 0.9210447      EUR  rs4409764             ref 0.4780</span></span>
<span><span class="co">#&gt; 4: 0.8099419 0.9216732      EUR  rs4409764             ref 0.4780</span></span>
<span><span class="co">#&gt; 5: 0.8134992 0.9613679      EUR  rs4409764             ref 0.4780</span></span>
<span><span class="co">#&gt; 6: 0.8102731 0.9219866      EUR rs10883365             ref 0.4797</span></span>
<span><span class="co">#&gt;                phecode_string category_string cat_LOG10_P   cat_OR cat_L95</span></span>
<span><span class="co">#&gt; 1:                   Migraine    neurological    9.522879 1.086957   1.060</span></span>
<span><span class="co">#&gt; 2:             Hypothyroidism       endocrine   10.000000 1.088000   1.060</span></span>
<span><span class="co">#&gt; 3: Inflammatory bowel disease       digestive   54.000000 1.182000   1.149</span></span>
<span><span class="co">#&gt; 4:         Regional enteritis       digestive   46.221849 1.188958   1.170</span></span>
<span><span class="co">#&gt; 5:         Ulcerative colitis       digestive   35.698970 1.171065   1.150</span></span>
<span><span class="co">#&gt; 6:         Regional enteritis       digestive    9.397940 1.180000   1.050</span></span>
<span><span class="co">#&gt;    cat_U95 Study_accession pub_count   pub_date powered rep     Power      rOR</span></span>
<span><span class="co">#&gt; 1:   1.120      GCST003720         1 2016-06-20       0   0 0.2360927 1.005937</span></span>
<span><span class="co">#&gt; 2:   1.116      GCST003988         1 2016-05-16       1   1 0.9446085 1.041313</span></span>
<span><span class="co">#&gt; 3:   1.217      GCST001725         1 2012-11-01       1   1 0.9966647 1.151809</span></span>
<span><span class="co">#&gt; 4:   1.210      GCST003044         2 2015-07-20       1   1 0.9980411 1.157403</span></span>
<span><span class="co">#&gt; 5:   1.200      GCST003045         1 2015-07-20       1   1 0.9098123 1.130776</span></span>
<span><span class="co">#&gt; 6:   1.320      GCST000039         1 2007-06-06       0   1 0.3249896 1.156970</span></span>
<span><span class="co">#&gt;         rL95     rU95 CI_overlap</span></span>
<span><span class="co">#&gt; 1: 0.9186078 1.101569    overlap</span></span>
<span><span class="co">#&gt; 2: 1.0074935 1.076269    overlap</span></span>
<span><span class="co">#&gt; 3: 1.0857236 1.221917    overlap</span></span>
<span><span class="co">#&gt; 4: 1.0849833 1.234656    overlap</span></span>
<span><span class="co">#&gt; 5: 1.0401845 1.229257    overlap</span></span>
<span><span class="co">#&gt; 6: 1.0846145 1.234152    overlap</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-the-replication-rate-of-associations-powered-at-80">Calculate the replication rate of associations powered at
&gt;80%<a class="anchor" aria-label="anchor" href="#calculate-the-replication-rate-of-associations-powered-at-80"></a>
</h2>
<p>The get_RR function takes an annotated data.table and calcuated the
powered replication rate.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_RR.html">get_RR</a></span><span class="op">(</span><span class="va">anno</span><span class="op">)</span></span>
<span><span class="co">#&gt; Replicated 651 of 853 for RR=76.3%</span></span>
<span><span class="co">#&gt; [1] 0.7631887</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-the-overall-replication-rate">Calculate the overall replication rate<a class="anchor" aria-label="anchor" href="#calculate-the-overall-replication-rate"></a>
</h2>
<p>By adding the include=‘all’ option, the replication rate is
calculated using all associations, regardless of power. This measure is
known as the overall replication rate.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_RR.html">get_RR</a></span><span class="op">(</span><span class="va">anno</span>,include<span class="op">=</span><span class="st">'all'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Replicated 1354 of 3268 for RR=41.4%</span></span>
<span><span class="co">#&gt; [1] 0.4143207</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="calculate-the-actualexpected-ratio">Calculate the actual:expected ratio<a class="anchor" aria-label="anchor" href="#calculate-the-actualexpected-ratio"></a>
</h2>
<p>The actual:expected ratio (AER) is the measure of the number of
associations replicated divided by the sum of the power. This measure
avoids the threshold effect induced by excluding on power &gt;80%, and
can be particularly useful for smaller datasets.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_AER.html">get_AER</a></span><span class="op">(</span><span class="va">anno</span>, LOUD <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Expected 1676.7, replicated 1354 for AE=0.808 (3268 associations for 106 uniq phecodes)</span></span>
<span><span class="co">#&gt; [1] 0.8075431</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="replication-by-phenotype-category">Replication by phenotype category<a class="anchor" aria-label="anchor" href="#replication-by-phenotype-category"></a>
</h2>
<p>You can assess the replication rate by category like this.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">by_pheno_cat</span><span class="op">=</span><span class="va">anno</span><span class="op">[</span>,<span class="fu">.</span><span class="op">(</span>total_tested <span class="op">=</span> <span class="va">.N</span>, replicated <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">rep</span><span class="op">)</span>, AER <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">rep</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">Power</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'category_string'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">by_pheno_cat</span></span>
<span><span class="co">#&gt;         category_string total_tested replicated       AER</span></span>
<span><span class="co">#&gt;  1:        neurological          234        103 0.8595045</span></span>
<span><span class="co">#&gt;  2:           endocrine          330        198 0.8737966</span></span>
<span><span class="co">#&gt;  3:           digestive          517        241 0.9160612</span></span>
<span><span class="co">#&gt;  4:           neoplasms          669        289 0.8912240</span></span>
<span><span class="co">#&gt;  5:         respiratory          292         92 0.5983357</span></span>
<span><span class="co">#&gt;  6:  circulatory system          402        199 0.8818822</span></span>
<span><span class="co">#&gt;  7:         psychiatric          331         24 0.4104517</span></span>
<span><span class="co">#&gt;  8:        sense organs          188         59 0.6048072</span></span>
<span><span class="co">#&gt;  9:      metabolic/heme           33         23 0.7728091</span></span>
<span><span class="co">#&gt; 10:        dermatologic          138         79 0.7380070</span></span>
<span><span class="co">#&gt; 11:     musculoskeletal           75         23 0.6221231</span></span>
<span><span class="co">#&gt; 12:       genitourinary           48         21 0.6895891</span></span>
<span><span class="co">#&gt; 13: infectious diseases           11          3 0.9277110</span></span></code></pre></div>
<p>Why not make a plot?</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">by_pheno_cat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">category_string</span>, x <span class="op">=</span> <span class="va">AER</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>stat <span class="op">=</span> <span class="st">'identity'</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span></span></code></pre></div>
<p><img src="my-vignette_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="comparing-a-single-phenotype">Comparing a single phenotype<a class="anchor" aria-label="anchor" href="#comparing-a-single-phenotype"></a>
</h2>
<p>You can also compare the associations for a single phenotype against
what was reported in the catalog. To do this, we use the rOR, rL95, and
rU95 values from the annotion file as columns have been oriented to the
risk allele. To make a plot, we will bind results from the test cohort
and the source cohort (catalog) for a single phenotype.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gout_test</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span><span class="va">phecode_string</span> <span class="op"><a href="https://Rdatatable.gitlab.io/data.table/reference/like.html" class="external-link">%like%</a></span> <span class="st">'Gout'</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="st">'rOR'</span>, <span class="st">'rL95'</span>, <span class="st">'rU95'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gout_test</span><span class="op">$</span><span class="va">cohort</span> <span class="op">=</span> <span class="st">'test'</span></span>
<span><span class="va">gout_cat</span> <span class="op">=</span> <span class="va">anno</span><span class="op">[</span><span class="va">phecode_string</span> <span class="op"><a href="https://Rdatatable.gitlab.io/data.table/reference/like.html" class="external-link">%like%</a></span> <span class="st">'Gout'</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SNP'</span>, <span class="st">'cat_OR'</span>, <span class="st">'cat_L95'</span>, <span class="st">'cat_U95'</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">gout_cat</span><span class="op">$</span><span class="va">cohort</span> <span class="op">=</span> <span class="st">'catalog'</span></span>
<span><span class="va">gout</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">gout_test</span>, <span class="va">gout_cat</span>, use.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gout</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">SNP</span>, x <span class="op">=</span> <span class="va">rOR</span>, color <span class="op">=</span> <span class="va">cohort</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_linerange.html" class="external-link">geom_errorbar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xmin <span class="op">=</span> <span class="va">rL95</span>, xmax <span class="op">=</span> <span class="va">rU95</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">0.2</span>, position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">'Odds ratio'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="my-vignette_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Lisa Bastarache, Jake Hughey.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
